### DISCUSSION TITLE
[Documentation] Add Step-by-Step Tutorial and Enhance Code Structure

### DISCUSSION THREAD

**Discussion by otaviojava** [oracle/nosql#23] [Q&A] [answered]
I've compared our NoSQL database SDK with others like Cassandra and MongoDB, and I believe we can improve our user experience. Here are the specific changes I'd like to propose:

**1. Step-by-Step Tutorial:**
- We need a comprehensive, step-by-step tutorial for newcomers who want to use your SDK. This Tutorial should include simple, easy-to-follow steps to achieve specific goals. For instance, a tutorial to set up and use our SDK with Docker might look like this:

   - Check if you have Docker installed and running.
   - Run the following command to start Oracle NoSQL: [Provide the command here]
   - Configure your environment settings.
   - Create a sample class or script to interact with our SDK.

**2. Exploring Multiple Sessions:**
- Currently, our SDK relies on a single, monolithic class for operations. It would be more user-friendly to break this down into multiple sessions or classes for each operation. This way, users can focus on their needed functionality without getting overwhelmed.

**References:**
- [Cassandra Java Driver Usage Examples](https://www.baeldung.com/cassandra-datastax-java-driver)
- [MongoDB Java Driver Usage Examples](https://www.mongodb.com/docs/drivers/java/sync/current/usage-examples/insertOne/)

Please prioritize these improvements to enhance the user experience for newcomers and make our SDK more competitive in the NoSQL database market.

**Comment by gmfeinberg** [comment]
Thanks for taking the time to think about this and comment. We'd love to improve the user experience. I will make sure that this gets to the relevant teams.

Can you please clarify your comment (2) about multiple sessions. I don't understand what you are suggesting

**Comment by otaviojava** [comment]
> Thanks for taking the time to think about this and comment. We'd love to improve the user experience. I will make sure that this gets to the relevant teams.
> 
> Can you please clarify your comment (2) about multiple sessions. I don't understand what you are suggesting

Yes, for example:

- Configuration (explain the possibilities of configurations)
- Data operations
- etc.


For example, the Apache Cassandra:

![Screenshot 2024-01-13 at 17 17 25](https://github.com/oracle/nosql-java-sdk/assets/863011/e3308672-1798-4e81-aa65-63d4c2da3ebe)

**Comment by gmfeinberg** [comment]
We have some of that for our other language SDKs but unfortunately not for Java at this time. Is this the sort of information you mean?
https://nosql-python-sdk.readthedocs.io/en/latest/

We've been planning to add something similar for Java using GitHub pages. Would that work?

**Comment by dario-vega** [comment, marked as answer]
Oi Otavio

In our NoSQL product documentation, We have **also** a document called [Oracle NoSQL Database - Developers Guide](https://docs.oracle.com/en/database/other-databases/nosql-database/23.3/nsdev/get-started.html) with a rich Step-by-Step Tutorial. The goal of this tutorial is to provide the best user experience for newcomers and make our SDK more competitive in the NoSQL database market. It uses a concept that doc writers call articles, hope it allows users to focus on their needed functionality without getting overwhelmed.

Of course, there are more details in the other documents of the NoSQL product documentation.

@Vandana-Rajamani is improving this documentation to add information about the SDKs, and AFAIU, this version is still under review and it will be released very soon. You can find the sample class that will be included in this documentation in our GitHub `nosql-examples` repository  [here](https://github.com/oracle/nosql-examples/tree/master/examples-nosql-java-sdk/sqlexamples)

Note: it will be done in a format similar to the documentation for the cloud - e.g [Obtaining a NoSQL Handle](https://docs.oracle.com/en/cloud/paas/nosql-cloud/onsql/#articletitle). As discussed the only difference between a program connected to the cloud and on-premise is the way to obtain this handle.

If needed, We can also create a blog and Tutorial as we did for the cloud but following the template of your examples. Using a deployment based on a container image.

- [LiveLabs: Get started with tables in Oracle NoSQL Database Cloud Service](https://apexapps.oracle.com/pls/apex/r/dbpm/livelabs/view-workshop?wid=642)
- [Blog: Oracle NoSQL Database Cloud Service – 15 Minutes to Hello World](https://blogs.oracle.com/developers/post/oracle-nosql-database-cloud-service-15-minutes-to-hello-world)

Finally, I will move this issue to a GitHub discussion in the nosql-examples GitHub repository. Documentation is an effort that we want to have for all NoSQL SDK

- [Oracle NoSQL SDK for Java](https://github.com/oracle/nosql-java-sdk)
- [Node.js for Oracle NoSQL Database](https://github.com/oracle/nosql-node-sdk)
- [.NET SDK for Oracle NoSQL Database](https://github.com/oracle/nosql-dotnet-sdk)
- [Oracle NoSQL Database Go SDK](https://github.com/oracle/nosql-go-sdk)
- [Oracle NoSQL Database SDK for Spring Data](https://github.com/oracle/nosql-spring-sdk)
- [Oracle NoSQL Database Python SDK](https://github.com/oracle/nosql-python-sdk)
 
Note: we have also something that we call Java Direct Driver included directly with the database - it is a low-level interface.  I don't recommend it for the usage that we are thinking of.

**Comment by dario-vega** [comment]
**ERRATUM**: the current version [Oracle NoSQL Database - Developers Guide](https://docs.oracle.com/en/database/other-databases/nosql-database/23.3/nsdev/get-started.html) is providing snippets for each SDK.

**Comment by Vandana-Rajamani** [comment]
Thanks @otaviojava otaviojava. A new version with improvements will be published soon!

**Comment by Vandana-Rajamani** [comment]
@otaviojava Hi Otavio, A new version of [Oracle NoSQL Database - Developers Guide](https://docs.oracle.com/en/database/other-databases/nosql-database/23.3/nsdev/index.html) with code snippets for NoSQL SDKs( Java,Python, Go, Javascript, Typescript, .NET) is available.
